File: bench/kernels/list_alloc/ref.c
/* list_alloc: build, sum, and free a 10k-node linked list per op.
 * The allocation-discipline kernel: malloc-per-node is the C baseline the
 * wolf region allocator (arena bump, wholesale free) intends to beat (D12).
 * Protocol: argv[1]=ops; prints {"ns":<total>,"ops":<ops>}. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define NODES 10000

struct node { uint64_t v; struct node *next; };

volatile uint64_t sink;

int main(int argc, char **argv) {
    uint64_t ops = argc > 1 ? strtoull(argv[1], 0, 10) : 100;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (uint64_t k = 0; k < ops; k++) {
        struct node *head = 0;
        for (int i = 0; i < NODES; i++) {
            struct node *n = malloc(sizeof *n);
            n->v = (uint64_t)i;
            n->next = head;
            head = n;
        }
        uint64_t sum = 0;
        for (struct node *n = head; n; n = n->next) sum += n->v;
        sink = sum;
        while (head) {
            struct node *next = head->next;
            free(head);
            head = next;
        }
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%llu}\n",
           (unsigned long long)ns, (unsigned long long)ops);
    return 0;
}

File: bench/kernels/list_alloc/ref.rs
//! list_alloc (family B), Rust reference: `Vec<Node>` is Rust's idiomatic
//! answer — one allocation, one free, which is closer to wolf's region
//! than to naive C's malloc-per-node.
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;
use std::time::Instant;

struct Node {
    v: i64,
    next: i64,
}

fn build_walk(nodes: i64) -> i64 {
    let mut xs: Vec<Node> = Vec::new();
    for i in 0..nodes {
        xs.push(Node {
            v: i & 1023,
            next: i - 1,
        });
    }
    let mut sum: i64 = 0;
    let mut idx = nodes - 1;
    while idx >= 0 {
        let n = &xs[idx as usize];
        sum = (sum + n.v) & 1048575;
        idx = n.next;
    }
    sum
}

fn main() {
    let ops: i64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(200);
    let nodes: i64 = 10000;
    let t = Instant::now();
    let mut sink: i64 = 0;
    for _ in 0..ops {
        sink = (sink + build_walk(black_box(nodes))) & 1048575;
    }
    let ns = t.elapsed().as_nanos();
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink}}}",
        ops.wrapping_mul(nodes)
    );
}
